File: lab2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"unsafe"
)

func format_input() ([]byte, int) {

	var result [100]byte

	//Ask for input and separate it in bytes
	reader := bufio.NewReader(os.Stdin)
	fmt.Print("Enter a string (Max 100 characters): ")
	raw_input, _ := reader.ReadString('\n')
	input_bytes := []byte(raw_input)

	//Inicialize pointers
	ptr_input := &input_bytes[0]
	ptr_result := &result[0]

	count := 0
	valid_count := 0

	for {
		//Character as the value of input pointer
		character := *ptr_input

		if character == '\n' || count >= 100 {
			break
		}

		if (character >= 'A' && character <= 'Z') || (character >= 'a' && character <= 'z') || (character == ' ' && valid_count > 0) {
			//Value of result pointer change to character
			*ptr_result = character
			//Step up result pointer
			ptr_result = (*byte)(unsafe.Pointer(uintptr(unsafe.Pointer(ptr_result)) + 1))
			valid_count++
		}

		count++

		//Step up input pointer
		ptr_input = (*byte)(unsafe.Pointer(uintptr(unsafe.Pointer(ptr_input)) + 1))
	}

	return result[:valid_count], valid_count

}

func reverse_list(letter_list *[]byte, length int) {

	if length <= 1 {
		return
	}

	//Pointers at start and end of the string
	left_ptr := &(*letter_list)[0]
	right_ptr := (*byte)(unsafe.Pointer(uintptr(unsafe.Pointer(left_ptr)) + uintptr(length-1)))

	count := 0

	for count < length/2 {

		*left_ptr, *right_ptr = *right_ptr, *left_ptr

		//For doing pointer arithmetic
		left_ptr = (*byte)(unsafe.Pointer(uintptr(unsafe.Pointer(left_ptr)) + 1))
		right_ptr = (*byte)(unsafe.Pointer(uintptr(unsafe.Pointer(right_ptr)) - 1))

		count++
	}

}

func count_vowels_consonants(letter_list *[]byte, length int) (map[byte]int, int, int) {

	vowels_counts := map[byte]int{}
	vowels_bytes := [5]byte{'A', 'E', 'I', 'O', 'U'}
	number_of_vowels := 0
	number_of_consonants := 0

	ptr_letter := &(*letter_list)[0]

	count_letter_list := 0

	//For to traverse letter list
	for {
		character := (*ptr_letter)
		ptr_vowels := &vowels_bytes[0]

		if count_letter_list == length {
			break
		}

		count_vowels_list := 0
		is_vowel := false

		//For to traverse and compare vowels
		for {
			vowel := (*ptr_vowels)

			if count_vowels_list == 5 {
				break
			}

			//Compare vowels in lower and upper
			if character == vowel || character == vowel+32 {

				vowels_counts[vowel+32] += 1
				number_of_vowels++
				is_vowel = true

				break
			}

			ptr_vowels = (*byte)(unsafe.Pointer(uintptr(unsafe.Pointer(ptr_vowels)) + 1))
			count_vowels_list++
		}

		if character != ' ' && !is_vowel {
			number_of_consonants++
		}

		ptr_letter = (*byte)(unsafe.Pointer(uintptr(unsafe.Pointer(ptr_letter)) + 1))
		count_letter_list++

	}

	return vowels_counts, number_of_vowels, number_of_consonants

}

func format_vowels_consonants_output(vowels, consonants int, type_vowels string) string {

	spaced := strings.Join(strings.Split(type_vowels, ""), " ")
	output := fmt.Sprintf(" %d %s %d ", vowels, spaced, consonants)

	return output
}

func sort_map(vowels_map map[byte]int) string {
	vowels_order := [5]byte{'a', 'e', 'i', 'o', 'u'}

	ptr := &vowels_order[0]
	count := 0

	output := ""

	for {

		if count == 5 {
			break
		}

		vowel := *ptr
		value := vowels_map[vowel]
		if value > 0 {
			output += fmt.Sprint(value)
		}

		ptr = (*byte)(unsafe.Pointer(uintptr(unsafe.Pointer(ptr)) + 1))
		count++

	}

	return output
}

func change_space_for_underscores(letter_list *[]byte, length int) {

	ptr_letter := &(*letter_list)[0]
	count := 0

	for {
		character := *ptr_letter

		if length == count {
			break
		}

		if character == ' ' {
			*ptr_letter = '_'
		}

		count++
		ptr_letter = (*byte)(unsafe.Pointer(uintptr(unsafe.Pointer(ptr_letter)) + 1))
	}

}

func main() {

	//Clean string input converting it in a slice
	input_list, length := format_input()

	reverse_list(&input_list, length)
	fmt.Print(string(input_list))

	vowels_counts, number_of_vowels, number_of_consonants := count_vowels_consonants(&input_list, length)
	ordered_vowels_count := sort_map(vowels_counts)
	fmt.Print(format_vowels_consonants_output(number_of_vowels, number_of_consonants, ordered_vowels_count))

	change_space_for_underscores(&input_list, length)
	fmt.Print(string(input_list))
}

File: lab2.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

void format_input(char *result, int *valid_count) {
    char input[200];
    printf("Enter a string (Max 100 characters): ");
    fgets(input, sizeof(input), stdin);

    int count = 0;
    *valid_count = 0;

    for (int i = 0; i < 100 && input[i] != '\0' && input[i] != '\n'; i++) {
        char character = input[i];
        if (isalpha(character) || (character == ' ' && *valid_count > 0)) {
            result[*valid_count] = character;
            (*valid_count)++;
        }
        else{
            break;
        }
    }
}

void reverse_list(char *list, int length) {
    for (int i = 0; i < length / 2; i++) {
        char temp = list[i];
        list[i] = list[length - i - 1];
        list[length - i - 1] = temp;
    }
}

void count_vowels_consonants(char *list, int length, int *vowel_counts, int *num_vowels, int *num_consonants) {
    char vowels[5] = {'A', 'E', 'I', 'O', 'U'};
    *num_vowels = 0;
    *num_consonants = 0;
    memset(vowel_counts, 0, 5 * sizeof(int));

    for (int i = 0; i < length; i++) {
        char c = toupper(list[i]);
        int is_vowel = 0;
        for (int j = 0; j < 5; j++) {
            if (c == vowels[j]) {
                vowel_counts[j]++;
                (*num_vowels)++;
                is_vowel = 1;
                break;
            }
        }
        if (!is_vowel && c != ' ') {
            (*num_consonants)++;
        }
    }
}

void format_vowels_consonants_output(int vowels, int consonants, int *vowel_counts, char *output) {
    sprintf(output, " %d ", vowels);
    for (int i = 0; i < 5; i++) {
        if (vowel_counts[i] > 0) {
            char temp[10];
            sprintf(temp, "%d ", vowel_counts[i]);
            strcat(output, temp);
        }
    }
    char consonant_str[10];
    sprintf(consonant_str, "%d ", consonants);
    strcat(output, consonant_str);
}

void change_space_for_underscores(char *list, int length) {
    for (int i = 0; i < length; i++) {
        if (list[i] == ' ') {
            list[i] = '_';
        }
    }
}

int main() {
    char result[100] = {0};
    int valid_count;

    format_input(result, &valid_count);

    reverse_list(result, valid_count);
    printf("%.*s", valid_count, result);

    int vowel_counts[5];
    int num_vowels, num_consonants;
    count_vowels_consonants(result, valid_count, vowel_counts, &num_vowels, &num_consonants);

    char formatted_output[100] = {0};
    format_vowels_consonants_output(num_vowels, num_consonants, vowel_counts, formatted_output);
    printf("%s", formatted_output);

    change_space_for_underscores(result, valid_count);
    printf("%.*s", valid_count, result);

    return 0;
}
